### PR TITLE
Solver history

### PR DESCRIPTION
**Pull request by tjhei**
- dump to file output/solver_history.txt if the solver fails
- add a signal/slot to access solver history data even if solver
succeeds
- add test

### CODE REVIEW

**Comment by tjhei** [comment]
(wait with merge until #595 is merged and I rebased)

File: source/simulator/solver.cc [hunk 1]
```diff
@@ -664,6 +682,8 @@ namespace aspect
     // the simple solver failed
     catch (SolverControl::NoConvergence)
       {
+        solver_history.push_back(-1);
```
Comment (bangerth) [inline source/simulator/solver.cc]: I think a comment would be useful here about what the intent of the line is.

File: source/simulator/solver.cc [hunk 2]
```diff
@@ -458,7 +458,15 @@ namespace aspect
   }
 
 
-
+  dealii::SolverControl::State
+  solver_callback (const unsigned int iteration,
```
Comment (bangerth) [inline source/simulator/solver.cc]: Put the function in an anonymous namespace. No need to exert the linker more than necessary.

File: source/simulator/solver.cc [hunk 3]
```diff
@@ -687,14 +713,32 @@ namespace aspect
...
+                      f << i << " " << solver_history[i] << "\n";
+                  }
+                f.close();
+
+                std::cout << "See " << parameters.output_directory+"solver_history.txt"
+                          << " for convergence history." << std::endl;
+
+                AssertThrow (false,
+                             ExcMessage (std::string("The iterative Stokes solver "
+                                                     "did not converge. It reported the following error:\n\n")
```
Comment (bangerth) [inline source/simulator/solver.cc]: Put the error text above into the error message of the assertion -- this way everything comes out in the same place.

**Comment by bangerth** [comment]
Please address my small niggles and merge yourself. Maybe an entry in the changelog?

**Comment by bangerth** [comment]
As I said above, please rebase and merge.

**Comment by bangerth** [comment]
github shows me that I can merge this. I think that's because the compilation fix commit is simply duplicated in whole for this PR, rather than being part of the patch. Will merge myself.

**Comment by tjhei** [comment]
Thanks. Yeah, git/github is pretty clever it looks like. :-)

**Pull request merged**